Class Definition and HBM for Holidays

nderguti opened the pull request:
https://synjira.atlassian.net/browse/NJPL-178

nderguti committed:
bankholidays crated and the pom.xml

nderguti committed:
Classes and Interface added components and the hbm.xml

ebeutler commented on bankholidays/pom.xml:
```diff
@@ -0,0 +1,50 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
```
@fpichler do we usually have a license notice in pom files?

fpichler replied:
yes. e.g. https://github.com/celements/celements-core/blob/dev/pom.xml

ebeutler commented on bankholidays/src/main/java/com/celements/calendar/bankholidays/classes/BankHolidayClass.java:
```diff
@@ -0,0 +1,54 @@
+  public static final String CLASS_SPACE = "Classes";
+  public static final String CLASS_DEF_HINT = CLASS_SPACE + "." + CLASS_NAME;
+
+  public static final ClassField<String> FIELD_HOLIDAY_NAME = new StringField.Builder(
```
please add a validation regexp and dictionary entry to prevent too long names that would overflow the DB field.
-> still missing. The added regexp checks for date, not name

ebeutler commented on bankholidays/src/main/resources/bankholidays_classes.hbm.xml:
```diff
@@ -0,0 +1,20 @@
+<hibernate-mapping>
+  <class entity-name="Classes.BankHolidayClass" table="cel_bankholidays_bank_holiday">
+    <id name="id" type="integer" unsaved-value="undefined">
+      <column name="CEC_ID" not-null="true" />
```
Why CEC? Usually we use something short to make field names recognisabe without knowing the table name. I would suggest BH (bank holiday) | CBH (celements bank holiday) | BHC (bank holiday class)

nderguti replied:
@ebeutler I used this because it was part of the Celements Calendar, I'll change it

ebeutler commented on bankholidays/src/main/resources/bankholidays_classes.hbm.xml:
```diff
@@ -0,0 +1,20 @@
+      <column name="CEC_HOLIDAY_TO" />
+    </property>
+  </class>
+</hibernate-mapping>
```
Missing new line to end the file

nderguti committed:
CBH (CELEMENTS BANK HOLIDAYS) added and the new line

nderguti committed:
Validation regex with time and the validation message

ebeutler commented on bankholidays/src/main/java/com/celements/calendar/bankholidays/classes/BankHolidayClass.java:
```diff
@@ -0,0 +1,55 @@
+          VALIDATION_REGEXP).validationMessage(VALIDATION_MESSAGE).size(30).build();
+
+  public static final ClassField<Date> FIELD_HOLIDAY_FROM = new DateField.Builder(CLASS_DEF_HINT,
+      "holiday_from").prettyName("Holiday from").dateFormat(DATE_FORMAT_TIMESTAMP).emptyIsToday(
```
validation regexp and message missing (as well as for "holiday to")

ebeutler commented on bankholidays/src/main/java/com/celements/calendar/bankholidays/classes/BankHolidayClassesRole.java:
```diff
@@ -0,0 +1,14 @@
+
+  public static final String DATE_FORMAT_TIMESTAMP = "dd.MM.yyyy HH:mm";
+  public static final String VALIDATION_REGEXP = "/^(0[1-9]|[12][0-9]|3[01])\\."
+      + "(0[1-9]|1[012])\\.(19|20|99)\\d\\d ([01]?[0-9]|2[0-3]):[0-5][0-9]$/";
```
According to the format hours is HH (not H). So the hours always have to be a two digit number. The ? is not needed.
[01]**_?_**[0-9]|2[0-3]

nderguti committed:
Refactored to FIELD_HOLIDAY_DATE 

The name and pretty name were not changed because in the ticket
(https://synjira.atlassian.net/browse/NJPL-178)
it is written thet I should have Angaben für "Feiertagsname":
Feiertag-Dokumenten müssen für jedes Jahr ein Objekt mit Angaben über
"Feiertagsname", "von" und "bis" Datum-Zeit erfasst werden können.

fpichler commented on bankholidays/src/main/resources/bankholidays_classes.hbm.xml:
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE hibernate-mapping PUBLIC "-//Hibernate/Hibernate Mapping DTD//EN"
+        "http://hibernate.sourceforge.net/hibernate-mapping-3.0.dtd">
+<hibernate-mapping>
+  <class entity-name="Classes.BankHolidayClass" table="cel_bankholidays_bank_holiday">
```
please use a module specific space name. e.g. CalendarClasses

fpichler commented on bankholidays/src/main/java/com/celements/calendar/bankholidays/classes/BankHolidayClass.java:
```diff
@@ -0,0 +1,55 @@
+public class BankHolidayClass extends AbstractClassDefinition implements BankHolidayClassesRole {
+
+  public static final String CLASS_NAME = "BankHolidayClass";
+  public static final String CLASS_SPACE = "Classes";
```
please use a module specific space name. e.g. CalendarClasses

nderguti committed:
Regular expression added for FIELD_HOLIDAY_NAME

nderguti committed:
Hour in regexp changed to two digits

nderguti committed:
CalendarClasses used instead of Classes

ebeutler commented on bankholidays/src/main/java/com/celements/calendar/bankholidays/classes/BankHolidayClassesRole.java:
```diff
@@ -0,0 +1,16 @@
+  public static final String VALIDATION_REGEXP_DATE_TIME = "/^(0[1-9]|[12][0-9]|3[01])\\."
+      + "(0[1-9]|1[012])\\.(19|20|99)\\d\\d (0[0-9]|1[0-9]|2[0-3]):[0-5][0-9]$/";
+  public static final String VALIDATION_MESSAGE_DATE_TIME = "bank_holiday_validation_date_time";
+  public static final String VALIDATION_REGEXP_HOLIDAY_NAME = "/^[a-zA-Z]{0,64}$/";
```
You can leave the details on the characters. I don't see a reason why we shouldn't allow names as e.g. "Christmas / Hanukkah". So just use "." instead of "[a-zA-Z]"

nderguti replied:
@ebeutler I changed it to:` "/^.{0,64}$/";` by using a dot "." so that every character can be taken but the problem is that any character means, for example `?!\ `these characters too!
Shouldn't this be just a Holiday Name, without characters ?

nderguti committed:
Allowing any character in VALIDATION_REGEXP_HOLIDAY_NAME

fpichler commented on bankholidays/src/main/resources/bankholidays_classes.hbm.xml:
```diff
@@ -0,0 +1,20 @@
+<!DOCTYPE hibernate-mapping PUBLIC "-//Hibernate/Hibernate Mapping DTD//EN"
+        "http://hibernate.sourceforge.net/hibernate-mapping-3.0.dtd">
+<hibernate-mapping>
+  <class entity-name="CalendarClasses.BankHolidayClass" table="cel_bankholidays_bank_holiday">
```
please change table name to `cel_bankholidays` the rest is just duplication.

fpichler reviewed (changes requested):
minor renaming

nderguti committed:
renamed table name
